component/encryptor/encryptor.go

```diff
@@ -0,0 +1,368 @@
+package encryptor
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"os"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/config"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+	"github.com/Azure/azure-storage-fuse/v2/internal/handlemap"
+)
+
+// Implements two mount approach with only one file for data and metadata.
+type Encryptor struct {
+	internal.BaseComponent
+	handle           *os.File
+	blockSize        uint64
+	mountPointCipher string
+	encryptionKey    []byte
+	lastChunkMeta    *LastChunkMeta
+}
+
+type LastChunkMeta struct {
+	sync.Mutex
+	farthestBlockSeen int64 // keep track of the farthest block that has been written to the file.
+	paddingLength     int64 // keep track of the padding length of the last block.
+}
+
+type encryptorOptions struct {
+	BlockSize          uint64 `config:"block-size-mb" yaml:"block-size-mb,omitempty"`
+	EncryptedMountPath string `config:"encrypted-mount-path" yaml:"encrypted-mount-path,omitempty"`
+	EncryptionKey      string `config:"encryption-key" yaml:"encryption-key,omitempty"`
+}
+
+const (
+	compName    = "encryptor"
+	AuthTagSize = 16
+	NonceSize   = 12
+	MetaSize    = 28 // Sum of AuthTagSize and NonceSize.
+)
+
+var _ internal.Component = &Encryptor{}
+
+func (e *Encryptor) Name() string {
+	return compName
+}
+
+func (e *Encryptor) SetName(name string) {
+	e.BaseComponent.SetName(name)
+}
+
+func (e *Encryptor) SetNextComponent(nc internal.Component) {
+	e.BaseComponent.SetNextComponent(nc)
+}
+
+func (e *Encryptor) Priority() internal.ComponentPriority {
+	return internal.EComponentPriority.LevelMid()
+}
+
+func (e *Encryptor) Configure(isParent bool) error {
+	log.Trace("Encryptor::Configure :  %s", e.Name())
+
+	conf := encryptorOptions{}
+	err := config.UnmarshalKey(e.Name(), &conf)
+	if err != nil {
+		log.Err("Encryptor::Configure : config error [invalid config attributes]")
+		return fmt.Errorf("error reading config for %s: %w", e.Name(), err)
+	}
+
+	key := os.Getenv("ENCRYPTION_KEY")
+	if key == "" {
+		key = conf.EncryptionKey
+		if key == "" {
+			log.Err("Encryptor::Configure : encryption key not set")
+			return fmt.Errorf("encryption key not set")
+		}
+	}
+
+	e.encryptionKey, err = base64.StdEncoding.DecodeString(key)
+	if err != nil {
+		log.Err("Encryptor::Configure : error decoding encryption key")
+		return fmt.Errorf("error decoding encryption key: %w", err)
+	}
+
+	e.blockSize = 1024 * 1024 // default block size is 1MB
```

Comment:
Can you set the defaults for all the options using viper ?

Reply:
I don't see viper.setDefault method implemented in config_parser.go file. And here in encryptor we don't have the hook for viper instance.